File: scala/trip_service_kata/src/main/scala/org/craftedsw/tripservicekata/trip/TripService.scala
package org.craftedsw.tripservicekata.trip

import org.craftedsw.tripservicekata.user.{UserSession, User}
import org.craftedsw.tripservicekata.exception.UserNotLoggedInException

import scala.util.control.Breaks._

class TripService(userSession: UserSession, tripDAO: TripDAO) {

	def getTripsOfFriend(user: User, loggedInUser: User): List[Trip] =
		user.friends().find(_ == loggedInUser).map(tripDAO.findTripsByUser).getOrElse(Nil)

	def getTripsByUser(user: User): List[Trip] =
		Option(userSession.getLoggedUser()).map(getTripsOfFriend(user, _)).getOrElse(throw new UserNotLoggedInException)

}

File: f#/TripServiceKata/TripServiceKata/TripService.fs
module TripServiceKata.Trip.TripService

open TripServiceKata
open TripServiceKata.User
open TripServiceKata.Trip.TripServiceDAO

exception UserNotLoggedInException

let GetTripsByUser (user: User): Trip list =
    let mutable tripList = []
    let loggedInUser = UserSession.GetLoggedUser
    let mutable friend = None
    if (Option.isSome loggedInUser)
    then
        for f in user.Friends do
            if f = Option.get loggedInUser
            then
                friend <- Some f
        if (Option.isSome friend)
        then
            tripList <- FindTripsByUser (Option.get friend)
        tripList
    else
        raise UserNotLoggedInException
